[framework] Added grid documentation (#864)
shopsys/shopsys: docs/cookbook/create-basic-grid.md

# Create Basic Grid
In this cookbook, we will create a new grid to display data.
We will learn how to properly use grid factory, how to configure data displayed, create a template to override some of default grid behaviors.
In the end, we will be able to even sort our grid by column, paginate large result sets and delete records.

## Prerequisites
This guide expects you to know how to work with controllers, its actions and you are familiar with concepts of administration pages in Shopsys Framework.
This guide is loosely linked to [Adding a New Administration Page](/docs/cookbook/adding-a-new-administration-page.md) cookbook and it is recommended to read it first.

## Create Grid factory
First, we need to create a factory responsible for the creation of our new grid.

> In `Shopsys\ShopBundle\Component\ArticleGrid` namespace we create the new class that will implement `GridFactoryInterface`.

Author: Oh, I noticed after I wrote the guide. Wanted to talk about it. For me, grid doesn't belong to the model. But we should be consistent.
I change it

            ->from(Article::class, 'a');

        return new QueryBuilderDataSource($queryBuilder, 'a.id');
    }
```

## Present data with Grid
We prepared our grid, but for now, it is not rendered anywhere and does not contain any columns.
We change this now.

First, we add columns we want to see into `create` method of `ArticleGridFactory`

```diff
    public function create(): Grid
    {
        $grid = $this->gridFactory->create('articleGrid', $this->createAndGetDataSource());

+       $grid->addColumn('id', 'a.id', t('Id'));
+       $grid->addColumn('name', 'a.name', t('Title'));
+       $grid->addColumn('createdAt', 'a.createdAt', t('Created'));

        return $grid;
    }
```

*Note: In the example above are column names translated. Do not forget to [dump translations](/docs/introduction/console-commands-for-application-management-phing-targets.md#dump-translations).*

Grid is ready to show all the articles from the database.
And we just need to render grid itself in a controller.
For this, we reuse the one created in [Adding a New Administration Page](/docs/cookbook/adding-a-new-administration-page.md) cookbook and add new action method for presenting articles.

> Do not forget to create a twig template for the new action.

Reviewer: I see you mention it below, however, I think it would be better to be more expressive so nobody omits the step :grin:

We need to inject (pass through constructor) `ArticleGridFactory` created earlier and pass grid view to the template.

```php
class DashboardController extends AdminBaseController
{

>     ...

Reviewer: I would expect here the added dependency on `ArticleGridFactory`, so new property and extended constructor :slightly_smiling_face:

    /**
     * @Route("/dashboard/articles/")
     */
    public function articlesAction()
    {

>         $grid = $this->awesomeGridFactory->create();

Reviewer: ```suggestion
        $grid = $this->articleGridFactory->create();
```

        return $this->render('@ShopsysShop/Admin/Content/Dashboard/articles.html.twig', [
            'gridView' => $grid->createView(),
        ]);
    }
```

> In the template, we render the whole grid just easy by calling `{{ gridView.render() }}` and grid will start do its work.

Reviewer: ```suggestion
In the template, we render the whole grid just easy by calling `{{ gridView.render() }}` and grid will do its work.
```

![Display Data with Grid](img/grid-display-data.png)

## Modify the basic grid appearance
As you probably noticed, dates in the third column are not printed much friendly.
To adjust appearance, we just need to extend the default grid template and modify it accordingly.

> We create the new twig template `articleGrid.html.twig` in `src/Shopsys/ShopBundle/Resources/views/Admin/Content`

Reviewer: ```suggestion
We create the new twig template `articleGrid.html.twig` in `src/Shopsys/ShopBundle/Resources/views/Admin/Content`.
```

> The template have to extend `@ShopsysFramework/Admin/Grid/Grid.html.twig` and override block `grid_value_cell_id_createdAt` <!--- TODO add after page will exist ([you can read more about blocks here](/docs/administration/grid-rendering.md)) -->

Reviewer: oh, I wanted to suggest you adding some explanation on how this works (something like `grid_value_cell_id_<columnId> where columnID is set as a 1st parameter in 
Grid::addColumn() method`) but ok, the TODO is fine for me now :grin:

```twig
{% extends '@ShopsysFramework/Admin/Grid/Grid.html.twig' %}

{% block grid_value_cell_id_createdAt %}

>     {{ value|formatDate }} {# pass value from the cell through helper #}

Reviewer: TBH, I do not understand the comment here

        ...

+       $grid->setTheme('@ShopsysShop/Admin/Content/Dashboard/articleGrid.html.twig');

        return $grid;
    }
```

## Sort rows
Now we see data as we want, but it would be nice to be able to adjust the view from the users perspective.
We may want to ease finding of a certain article by allowing to sort rows by name or date of creation.

A default order will be by date with newest articles at the top.

```diff
    public function create(): Grid
    {
        $grid = $this->gridFactory->create('articleGrid', $this->createAndGetDataSource());

        $grid->addColumn('id', 'a.id', t('Id'));
-       $grid->addColumn('name', 'a.name', t('Title'));
+       $grid->addColumn('name', 'a.name', t('Title'), true);
-       $grid->addColumn('createdAt', 'a.createdAt', t('Created'));
+       $grid->addColumn('createdAt', 'a.createdAt', t('Created'), true);

+       $grid->setDefaultOrder('createdAt', DataSourceInterface::ORDER_DESC);
    ...
```

## Paginate results
As the number of articles grow, the clarity decreases rapidly.
To keep work with grid enjoyable, we can split data across several pages with pagination.

> In the grid is just a matter of calling one method.

Reviewer: Hm, a simple question might come to the mind of a reader here, how many rows are displayed on a single page? And am I able to change this? :innocent:

```diff
    public function create(): Grid
    {
        $grid = $this->gridFactory->create('articleGrid', $this->createAndGetDataSource());

+       $grid->enablePaging();

        ...
```

> ## Delete articles from grid

Reviewer: The original title seems a little bit confusing - are you removing the articles from the view only or removing them from persistence?
```suggestion
## Delete articles using grid
```

> As the last, we may want to delete some articles.

Reviewer: ```suggestion
As the last, an admin may want to delete some articles.
```

Grid eases the task with already implemented action column.

We just use `addDeleteActionColumn` in existing `ArticleGridFactory` with arguments `admin_dashboard_delete` as a route (to handle request – we create proper action in a moment) and request parameters (action have to know id of the article to delete).
To prevent accidental deletion we can also set a confirmation message.

```diff
    public function create(): Grid
    {
        $grid = $this->gridFactory->create('articleGrid', $this->createAndGetDataSource());

        $grid->addColumn('id', 'a.id', t('Id'));
        $grid->addColumn('name', 'a.name', t('Title'));
        $grid->addColumn('createdAt', 'a.createdAt', t('Created'));

+       $grid->addDeleteActionColumn('admin_dashboard_delete', ['id' => 'a.id'])
+           ->setConfirmMessage(t('Really delete?'));
    ...
```

And now we add the action to delete the article.

```php
...
use Shopsys\FrameworkBundle\Component\Router\Security\Annotation\CsrfProtection;

class DashboardController extends AdminBaseController
{
    ...

    /**
     * @Route("/dashboard/articles/delete/{id}", requirements={"id" = "\d+"})

>      * @CsrfProtection

Reviewer: I am not sure whether this is already done in the linked cookbook where the DashboardController is created, however, you need to import `use Shopsys\FrameworkBundle\Component\Router\Security\Annotation\CsrfProtection` to make the annotation work properly :slightly_smiling_face:

Author: Actually, use statement is just a few lines above :)

     * @param int $id
     * @return \Symfony\Component\HttpFoundation\Response
     */
    public function deleteAction(int $id): Response
    {

>         // delete article

Reviewer: this won't be implemented here? It is really easy as there is a method prepared for that in `ArticleFacade` :grin:

# Create Basic Grid
In this cookbook, we will create a new grid to display salesmen data in the administration.
We will learn how to properly use grid factory, how to configure data displayed, and override the default data presentation.
In the end, we will be able to even sort our grid by column, paginate large result sets and delete records.

## Prerequisites

> * Provided we are going to list the salesmen, you need to create the Salesman entity and add some entries to the database table following [the cookbook](/docs/cookbook/adding-a-new-entity.md), first.

Reviewer: > was this result of grammarly ?
Provided we are going to === In the case we want to ??? BUT OK

No, this is my rich English :smile: I thought it is a synonym for "Because", however, it seems that it is not...so fixed

> or prerequisites is section where there is you pronoun ?

tbh, I was not paying attention to this kind of styling so it might be inconsistent...

* You should know how to work with controllers, its actions and be familiar with concepts of administration pages in Shopsys Framework
    * We recommended you to read [Adding a New Administration Page](/docs/cookbook/adding-a-new-administration-page.md) cookbook.

## 1. Define the grid

> ### 1.1 Create grid factory

Reviewer: as stated before, it seems more readable to me this way :slightly_smiling_face: I was getting lost all the time when there were no step numbers

```yaml
    Shopsys\ShopBundle\Grid\Salesman\SalesmanGridFactory: ~
```

### 1.2 Configure grid data source
`GridFactory::create()` requires an implementation of `DataSourceInterface` as its second argument.
So we create one that returns all salesmen we need.

In our `SalesmanGridFactory`, we add a new protected method, that creates and returns data source.
And because we want to get data from the database, we use data source created from Doctrine Query Builder.

```diff
+ use Doctrine\ORM\EntityManagerInterface;

+   /**
+    * @var \Doctrine\ORM\EntityManagerInterface
+    */
+   protected $entityManager;

-   public function __construct(GridFactory $gridFactory)
+   public function __construct(GridFactory $gridFactory, EntityManagerInterface $entityManager)
    {
        $this->gridFactory = $gridFactory;
+       $this->entityManager = $entityManager;
    }

    /**
     * @return \Shopsys\FrameworkBundle\Component\Grid\Grid
     */
    public function create(): Grid
    {
-       /* @TODO: implement */
+       $grid = $this->gridFactory->create('salesmanGrid', $this->createAndGetDataSource());

+       return $grid;
    }
```

> Now, let's implement `createAndGetDataSource` method that should be in the same `SalesmanGridFactory` and will look like this

Reviewer: Another thing that should be resolved for the whole knowledge base...

    {
        $grid = $this->gridFactory->create('salesmanGrid', $this->createAndGetDataSource());

+       $grid->addColumn('id', 's.id', t('Id'));
+       $grid->addColumn('name', 's.name', t('Name'));
+       $grid->addColumn('registeredAt', 's.registeredAt', t('Registered'));

        return $grid;
    }
```

*Note: In the example above, the column names are translated. Do not forget to [dump translations](/docs/introduction/console-commands-for-application-management-phing-targets.md#dump-translations).*

## 2. Display the grid

Grid is ready to show all the salesmen from the database.
And we just need to render grid itself using a controller.

### 2.1 Create a new controller & action
We need to inject (pass through constructor) `SalesmanGridFactory` created earlier and pass grid view to the template.
```php
namespace Shopsys\ShopBundle\Controller\Admin;

use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Shopsys\FrameworkBundle\Controller\Admin\AdminBaseController;
use Shopsys\ShopBundle\Grid\Salesman\SalesmanGridFactory;

class SalesmanController extends AdminBaseController
{
    /**
     * @var \Shopsys\ShopBundle\Grid\Salesman\SalesmanGridFactory
     */

>     protected $salesmanGridFactory;

Reviewer: oh, my bad, it should be there, of course, fixed.

    /**
     * @Route("/salesman/list/")
     */
    public function listAction()
    {
        $grid = $this->salesmanGridFactory->create();

        return $this->render('@ShopsysShop/Admin/Content/Salesman/list.html.twig', [
            'gridView' => $grid->createView(),
        ]);
    }
```

### 2.2 Create a new template

> Finally, it is time to create a new twig template, `src/Shopsys/ShopBundle/Resources/views/Admin/Content/Salesman/list.html.twig` with the following content:

Reviewer 2: on [shopsys/demoshop] we have it in `app/Resources` but we will se during tests

Reviewer: AFAIK, you need to have it there when you are overriding a template from the framework. But in this case, we are adding new functionality, so it should not matter where the templates are, right?

        ...
   }
```

## 5. Paginate results
As the number of salesmen grow, the clarity decreases rapidly.
To keep work with grid enjoyable, we can split data across several pages with pagination.

In the grid is just a matter of calling one method.

```diff
// src/Shopsys/ShopBundle/Grid/Salesman/SalesmanGridFactory

    public function create(): Grid
    {
        $grid = $this->gridFactory->create('salesmanGrid', $this->createAndGetDataSource());
        ...

+       $grid->enablePaging();

        ...
     }
```

## 6. Delete salesmen using the grid
As the last, an admin may want to delete some salesmen.
Grid eases the task with already implemented action column.

### 6.1 Implement the deletion logic
First, we need to get the salesman entity by its id and then remove it from persistence using Doctrine entity manager.
We will follow the basic concepts of Shopsys Framework (see ["Basics about model architecture"](/docs/introduction/basics-about-model-architecture.md) article) and create new classes - `SalesmanFacade` and `SalesmanRepository`.

#### 6.1.1 Create `SalesmanRepository` and implement `getById` method
 ```php

>  // src\Shopsys\ShopBundle\Model\Salesman\SalesmanRepository.php

Reviewer: we are not unified in this, I know...it needs to be reviewed globally I think :slightly_smiling_face:

    }
}
```

#### 6.1.2 Create `SalesmanFacade` and implement `deleteById` method
```php
// src\Shopsys\ShopBundle\Model\Salesman\SalesmanFacade.php
declare(strict_types=1);

namespace Shopsys\ShopBundle\Model\Salesman;

use Doctrine\ORM\EntityManagerInterface;

class SalesmanFacade
{
    /**
     * @var \Doctrine\ORM\EntityManagerInterface
     */
    protected $entityManager;

    /**
     * @var \Shopsys\ShopBundle\Model\Salesman\SalesmanRepository
     */
    protected $salesmanRepository;

    /**
     * @param \Doctrine\ORM\EntityManagerInterface $entityManager
     * @param \Shopsys\ShopBundle\Model\Salesman\SalesmanRepository $salesmanRepository
     */
    public function __construct(EntityManagerInterface $entityManager, SalesmanRepository $salesmanRepository)
    {
        $this->entityManager = $entityManager;
        $this->salesmanRepository = $salesmanRepository;
    }

    /**
     * @param int $salesmanId
     */
    public function deleteById(int $salesmanId)
    {
        $salesman = $this->salesmanRepository->getById($salesmanId);

        $this->entityManager->remove($salesman);
        $this->entityManager->flush();
    }
}
```

> ### 6.2 Add delete action to the controller

Reviewer: not sure about that, it does not sound correct to me :smile:

In our `SalesmanController`, we add the new action to delete the salesman using `SalesmanFacade`.
```diff
// src/Shopsys/ShopBundle/Controller/Admin/SalesmanController
...
+ use Shopsys\FrameworkBundle\Component\Router\Security\Annotation\CsrfProtection;
+ use Shopsys\ShopBundle\Model\Salesman\SalesmanFacade;

class SalesmanController extends AdminBaseController
{
+    /**
+     * @var \Shopsys\ShopBundle\Model\Salesman\SalesmanFacade
+     */
+    protected $salesmanFacade;

-    public function __construct(SalesmanGridFactory $salesmanGridFactory)
+    public function __construct(SalesmanGridFactory $salesmanGridFactory, SalesmanFacade $salesmanFacade)
     {
         $this->salesmanGridFactory = $salesmanGridFactory;
+        $this->salesmanFacade = $salesmanFacade;
+    }

+    /**

> +     * @Route("/dashboard/salesmen/delete/{id}", requirements={"id" = "\d+"})

Reviewer 2: `/dashboard/` ?

Reviewer 2: try this thing that works also for me
```php

    /**
     * @Route("/salesman/delete/{id}", requirements={"id" = "\d+"})
     * @CsrfProtection
     */
    public function deleteAction(int $id)
    {
        $this->salesmanFacade->deleteById($id);

        $this->getFlashMessageSender()
            ->addInfoFlash(t('Salesman with id %id% deleted', ['%id%' => $id]));

        return $this->redirectToRoute('admin_salesman_list');
    }
```

Reviewer: thanks, that was a leftover after the first implementation iteration :slightly_smiling_face:

### 3.1 Create a new template
We create the new twig template `listGrid.html.twig` in `src/Shopsys/ShopBundle/Resources/views/Admin/Content/Salesman`.
The template has to extend `@ShopsysFramework/Admin/Grid/Grid.html.twig` and override block `grid_value_cell_id_registeredAt` where we apply a Twig filter to the value.

> [You can read more about blocks here](/docs/administration/grid-rendering-customization.md).

Reviewer: yes, as I said, the magic with time traveling made this quite difficult :grin:
